src/Microsoft.AspNetCore.Mvc.Razor/TagHelpers/TagHelperComponentTagHelper.cs

```diff
@@ -20,29 +24,29 @@ public abstract class TagHelperComponentTagHelper : TagHelper
         /// <summary>
         /// Creates a new <see cref="TagHelperComponentTagHelper"/>.
         /// </summary>
-        /// <param name="components">The list of <see cref="ITagHelperComponent"/>.</param>
+        /// <param name="manager">The <see cref="ITagHelperComponentManager"/> which contains the list
+        /// of <see cref="ITagHelperComponent"/>s.</param>
         /// <param name="loggerFactory">The <see cref="ILoggerFactory"/>.</param>
-        public TagHelperComponentTagHelper(IEnumerable<ITagHelperComponent> components,
+        public TagHelperComponentTagHelper(ITagHelperComponentManager manager,
             ILoggerFactory loggerFactory)
         {
-            if (components == null)
+            if (manager == null)
             {
-                throw new ArgumentNullException(nameof(components));
+                throw new ArgumentNullException(nameof(manager));
             }
 
             if (loggerFactory == null)
             {
                 throw new ArgumentNullException(nameof(loggerFactory));
             }
 
-            _components = components;
+            _components = manager.Components.OrderBy(p => p.Order).ToArray();
```

Comment:
Add a comment mentioning this ordering must be done after the manager is created (otherwise controller-added components won't be positioned correctly).